text/0268-acceptance-testing-refactor.md

```diff
@@ -0,0 +1,304 @@
+  */
+  tap(selector: string | HTMLElement): Promise<void>;
+
+  /**
+    Triggers a keyboad event on the specified selector.
+  */
+  triggerKeyEvent(
+    selector: string | HTMLElement,
+    eventType: 'keydown' | 'keypress' | 'keyup',
+    keyCode: string,
+    modifiers?: {
+      ctrlKey: false,
+      altKey: false,
+      shiftKey: false,
+      metaKey: false
+    }
+  ): Promise<void>;
+
+  /**
+    Triggers an event on the specified selector.
+  */
+  triggerEvent(
+    selector: string | HTMLElement,
+    eventType: string,
+    eventOptions: any
+  ): Promise<void>;
+
+  /**
+    Fill in the specified selector's `value` property with the provided text.
+  */
+  fillIn(selector: string | HTMLElement, text: string): Promise<void>;
+
+  /**
+    Focus the specified selector.
+  */
+  focus(selector: string | HTMLElement): Promise<void>;
+
+  /**
+    Unfocus the specified selector.
+  */
+  blur(selector: string | HTMLElement): Promise<void>;
+
+  /**
+    Returns a promise which resolves when the provided callback returns a truthy value.
+  */
+  waitUntil(() => boolean): Promise<void>;
+
+  /**
+    Returns a promise which resolves when the provided selector (and count) becomes present.
+  */
+  waitFor(selector: string, count?: number): Promise<void>;
+}
+```
+
+### `setupAcceptanceTest`
+
+This function will:
+
+* invoke `ember-test-helper`s `setupContext` with the tests context (which does the following):
+  * create an owner object and set it on the test context (e.g. `this.owner`)
+  * setup `this.set`, `this.setProperties`, `this.get`, and `this.getProperties` to
+    the test context
+  * setup `this.pauseTest` and `this.resumeTest` methods to allow easy pausing/resuming
+    of tests
+* add routing related helpers
+  * setup `this.visit` method to visit the given url
+  * setup getter for `this.currentRouteName` which returns the current route name
+  * setup getter for `this.currentURL` which returns the current URL
+* add DOM interaction helpers (heavily influenced by @cibernox's lovely addon [ember-native-dom-helpers](https://github.com/cibernox/ember-native-dom-helpers))
+  * setup a getter for `this.element` which returns the DOM element representing
+    the applications root element
+  * setup `this.click` helper method
+  * setup `this.tap` helper method
+  * setup `this.triggerKeyEvent` helper method
+  * setup `this.triggerEvent` helper method
+  * setup `this.fillIn` helper method
+  * setup `this.focus` helper method
+  * setup `this.blur` helper method
+  * setup `this.waitUntil` helper method
+  * setup `this.waitFor` helper method
+
+### `setupRenderingTest`
+
+The `setupRenderingTest` function proposed in
+[emberjs/rfcs#232](https://github.com/emberjs/rfcs/blob/master/text/0232-simplify-qunit-testing-api.md)
+(and implemented in
+[ember-qunit](https://github.com/emberjs/ember-qunit)@3.0.0) will be modified to add the same DOM interaction helpers mentioned above:
+
+* setup `this.click` helper method
```

Comment:
I still find weird that we have two supported ways of doing the same thing. Also, specially in integration tests, it would be extraordinary that someone could do 
```
this.click = function(e) {
  assert.ok(e, 'The event is fired')
}
await render(hbs`{{my-component action=click}}`)
```

Do you think that supposed improvement in debuggeablility justifies this duality. Couldn't just the docs serve as ... documentation of what helpers exists and how to use them?

Reply:
FYI There is one potential advantage of using `await this.click()` over `await click()` in that you could potentially run tests concurrently. The `click()` variant has no reliable way of figuring out which test just called that helper, while `this.click()` is bound to the right context.

Reply:
> There is one potential advantage of using await this.click() over await click() in that you could potentially run tests concurrently.

Confirm, however I tried to avoid talking about this in #119 (and here) because there are many other issues with running tests in parallel (there is only one DOM, focus and other events don't fire the same, etc) and I didn't want to open a can of worms.

Reply:
Even if I slightly favour the imported helpers, my point is about having both. I think it should be one or the other.

Reply:
> Do you think that supposed improvement in debuggeablility justifies this duality? Couldn't just the docs serve as ...well... documentation of what helpers exists and how to use them?

Sure, I agree that this is definitely a good suggestion also. I'll add a fleshed out version of it to alternatives.

Reply:
> Even if I slightly favour the imported helpers, my point is about having both. I think it should be one or the other.
  
Right, totally understood. I guess I'm saying that the fact that there are both is an implementation detail. The primary public API and what blueprints will use will be the importable versions. The fact that they desugar to invoking methods on the test context is a minor detail (and _might_ enable features like what @Turbo87 alludes to in https://github.com/emberjs/rfcs/pull/268#discussion_r149123309).

Reply:
Agree. Made the same point [here](https://github.com/emberjs/rfcs/pull/268#discussion_r149086095). @rwjblue just added some clarification [here](https://github.com/emberjs/rfcs/pull/268/commits/4bb68572760b76d94b19809d0da84d186e05ec8b), which seems reasonable. I would not see it as a problem to have both, as long as there is *one* ideomatic way to write your tests, that is consistently used (guides, blueprints, codemods etc.)

Reply:
Oh, y'all did it now.  Just removed the `this.foo` versions from the RFC.  I agree having two different public API's for the same thing is confusing, and reserving a bunch of names on the test context is annoying and error prone for users. 

It is still likely that during implementation I will have _some_ mechanism to discover and access these helpers on the test context but it does not need to be public API at all...